### PR TITLE
[booster] support torch fsdp plugin in booster

### PR DESCRIPTION
**Pull request by wukong1992**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Booster support torch FSDP plugin.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                                state_dict_type: StateDictType, 
+                                state_dict_config: Optional[StateDictConfig],
+                                optim_state_dict_config: Optional[OptimStateDictConfig] = None,
+                                ) -> StateDictSettings:
+        return FSDP.set_state_dict_type(model,
+                                        state_dict_type,
+                                        state_dict_config,
+                                        optim_state_dict_config)
+
+    def load_unsharded_model(self, model: nn.Module, checkpoint: str):
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:61]: Will this support `save_sharded_model()`?

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                               forward_prefetch=forward_prefetch,
+                               limit_all_gathers=limit_all_gathers,
+                               use_orig_params=use_orig_params,
+                               ignored_parameters=ignored_parameters)
+
+    def support_no_sync(self) -> bool:
+        return True
+
+    def control_precision(self) -> bool:
+        return False
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:170]: I notice that there is a `mixed_precision` argument of this class. Are you sure that this class does not control precision?

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        self.__set_model_optim_state(model, 
+                                     StateDictType.LOCAL_STATE_DICT,
+                                     state_dict_config=LocalStateDictConfig())
+        model_state_dict = model.state_dict()
+        self.save_checkpoint(model_state_dict, checkpoint)
+
+    def save_sharded_optimizer(self, model: nn.Module, optimizer: Optimizer, checkpoint: str):
+
+        self.__set_model_optim_state(model, 
+                                     StateDictType.LOCAL_STATE_DICT,
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:91]: Did you check the shard method is the same as ours?

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                               ignored_parameters=ignored_parameters)
+
+    def support_no_sync(self) -> bool:
+        return True
+
+    def control_precision(self) -> bool:
+        return True
+
+    def supported_precisions(self) -> List[str]:
+        return ['fp16', 'fp16_apex', 'bf16', 'fp8']
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:208]: As FSDP controls precision itself, I think it cannot support "fp16_apex". Besides, torch doesn't support fp8 natively (see https://pytorch.org/docs/main/tensor_attributes.html?highlight=dtype#torch.dtype). I think this plugin also doesn't support "fp8".

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                               param_init_fn=param_init_fn,
+                               device_id=device_id,
+                               sync_module_states=sync_module_states,
+                               forward_prefetch=forward_prefetch,
+                               limit_all_gathers=limit_all_gathers,
+                               use_orig_params=use_orig_params,
+                               ignored_parameters=ignored_parameters)
+
+    def support_no_sync(self) -> bool:
+        return True
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:202]: If it supports `no_sync`, please

**Review by ver217 — APPROVED** [review]

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        >>> booster = Booster(plugin=plugin)
+        >>> model, optimizer, train_dataloader, criterion = booster.boost(model, optimizer, train_dataloader, criterion)
+
+    Args:
+        See https://pytorch.org/docs/stable/fsdp.html for details.
+    """
+
+    def __init__(self,
+                 *args,
+                 **kwargs,
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:178]: Why do you use `**kwargs`? It's hard for users to configure.

Comment (wukong1992) [reply colossalai/booster/plugin/torch_fsdp_plugin.py:178]: The init func of different versions of FSDP have different signatures, the number and types of parameters are different.

**Review by ver217 — APPROVED** [review]

**Comment by ver217** [comment]
> The code coverage for the changed files is 42%.
> 
> Click me to view the complete report

Torch version of CI is 1.11.0. Test is skipped.

**Pull request merged**